src/models/CDD.java

```diff
@@ -292,62 +329,69 @@ public CDD applyReset(List<Update> list) {
     }
 
     public boolean canDelayIndefinitely() {
-        if (isTrue()) {
-            return true;
-        }
-        if (isFalse()) {
-            return false;
-        }
-        if (isBDD()) {
-            return true;
+        if (!isGuardDirty) {
+            return canDelayIndefinitely;
         }
 
-        CDD copy = hardCopy();
-
-        while (!copy.isTerminal()) {
-            CddExtractionResult extraction = copy.removeNegative().reduce().extract();
-            copy = extraction.getCddPart().removeNegative().reduce();
-            Zone zone = new Zone(extraction.getDbm());
+        boolean result = true;
 
-            if (!zone.canDelayIndefinitely()) {
-                return false;
+        if (isFalse()) {
+            result = false;
+        } else if (!isBDD()) {
+            CDD copy = hardCopy();
+
+            while (!copy.isTerminal()) {
+                CddExtractionResult extraction = copy.removeNegative().reduce().extract();
+                copy = extraction.getCddPart().removeNegative().reduce();
+                Zone zone = new Zone(extraction.getDbm());
+
+                if (!zone.canDelayIndefinitely()) {
+                    result = false;
+                    break;
+                }
             }
         }
-        // found no states that cannot delay indefinitely
-        return true;
+
+        canDelayIndefinitely = result;
+
+        return result;
     }
 
     public boolean isUrgent() {
-        if (isTrue()) {
-            return false;
-        }
-        if (isFalse()) {
-            return true;
-        }
-        if (isBDD()) {
-            return false;
+        if (!isGuardDirty) {
+            return isUrgent;
         }
 
-        // Required as we don't want to alter the pointer value of "this"
-        CDD copy = hardCopy();
+        boolean result = !isBDD();
 
-        while (!copy.isTerminal()) {
-            CddExtractionResult res = copy.removeNegative().reduce().extract();
-            Zone zone = new Zone(res.getDbm());
-            copy = res.getCddPart().removeNegative().reduce();
-            if (!zone.isUrgent()) {
-                return false;
+        if (isFalse()) {
+            result = true;
+        } else {
+            // Required as we don't want to alter the pointer value of "this".
```

Comment:
Add this requirement also to `canDelayIndefinitely()`.

Reply:
I think `canDelayIndefinitely()` already makes a hard copy. But I might be missing something.

Reply:
Yes, but it does it in exactly the same place in the function. So that's why I was thinking of adding this comment to `canDelayIndefinitely()` as well.

Reply:
Good pointe. It has been added.